src/agent/src/main.rs

```diff
@@ -429,32 +436,43 @@ fn init_attestation_components(logger: &Logger, _config: &AgentConfig) -> Result
     )
     .map_err(|e| anyhow!("launch_process {} failed: {:?}", AA_PATH, e))?;
 
-    if let Err(e) = launch_process(
+    // skip launch of confidential-data-hub and api-server-rest
+    if config.guest_components_procs == GuestComponentsProcs::AttestationAgent {
+        return Ok(());
+    }
+
+    debug!(
+        logger,
+        "spawning confidential-data-hub process {}", CDH_PATH
+    );
+    launch_process(
         logger,
         CDH_PATH,
         &vec![],
         CDH_SOCKET,
         DEFAULT_LAUNCH_PROCESS_TIMEOUT,
-    ) {
-        error!(logger, "launch_process {} failed: {:?}", CDH_PATH, e);
-    } else {
-        let features = _config.guest_components_rest_api;
-        match features {
-            GuestComponentsFeatures::None => {}
-            _ => {
-                if let Err(e) = launch_process(
-                    logger,
-                    API_SERVER_PATH,
-                    &vec!["--features", &features.to_string()],
-                    "",
-                    0,
-                ) {
-                    error!(logger, "launch_process {} failed: {:?}", API_SERVER_PATH, e);
-                }
-            }
-        }
+    )
+    .map_err(|e| anyhow!("launch_process {} failed: {:?}", CDH_PATH, e))?;
```

Comment:
The previous logic was that if cdh/asr failed to launch, we would log the error, but then continue, but now as the config is explicitly staring which of them to run we throw an error to better match the user's expectations?

Reply:
oh, right. that's indeed a change of behaviour. I'm not sure if this was intentional, because we return an err if we fail to launch the AA process. Is there a reason why we would deem AA to be critical but tolerate a CDH failure?

Reply:
Yeah - I suspect that it's a bit of legacy build up in that the AA was there first and then CDH and ASR added. I think we had a discussion when we moved it over to `main` and the feeling was that the AA was a show stopper, but if the CDH didn't work then we could try and carry on. I'll see if I can find the thread and link it.

Reply:
https://github.com/kata-containers/kata-containers/pull/8870#discussion_r1550045052 might have been what I was remembering, but doesn't seem like a particularly concrete decision, so I'm fine with your implementation.

Reply:
hmm, yeah, I guess you can still use AA standalone and it seems a bit odd to attempt but tolerate a failed startup (it helps with debugging, but it's also a source of confusion).

I understood that image-rs will rely on CDH for layer decryption eventually, but it doesn't hurt to keep the prior behaviour, i.e. only abort when AA launch fails, and tolerate the others. WDYT?

Reply:
I guess my concern is that if a use specifies `agent.guest_components_procs=api-server-rest`, but then we don't launch CDH/ASR due to a failure in that launch, then I think the "obvious" user experience would be to fail, rather than log and proceed. The complication is that `agent.guest_components_procs=api-server-rest` is the default, so the user might not actually care about it in which case proceeding is fine! 😿

I guess if we might flip the CDH interaction then it's not super important to get this perfect now.

@fitzthum might be interested and have some thoughts on this and the other comment thread too?

Reply:
> I think the "obvious" user experience would be to fail, rather than log and proceed. The complication is that agent.guest_components_procs=api-server-rest is the default

Indeed, that's a bit messed up and hard to reconcile. I assume the intuitive green-field ux would be: A user requests the agent to launch a process (explicitly or implicity) and the agent itself fails hard if it cannot.

It is tough to debug in cases of failure, but I'm not sure this should be a guiding principle. We probably want to fail fast if anything attestation-related doesn't check out. And if we're confident about guest-components as a part of the CoCo stack, we shouldn't ignore breakage (the peerpod and kata CI is covering CDH + ASR now, afaik).

For debugging a developer could set `gc_procs=none` and attempt to launch the procs manually.

All that considered, I would recommend to fail in all cases.

Reply:
> All that considered, I would recommend to fail in all cases.

Yeah - I think that's fine

Reply:
Yeah this is an interesting one. Keep in mind that we already have runtime configurations that enable various feature that depend on these components. For instance if you enable the spi-server-rest, you need the whole stack. Other features (will) only require the CDH/AA. I am not against adding a config to explicitly set which processes are running, but it makes the user have to do a little bit more work.

Generally my instinct with this sort of thing is to fail fast and fatally, but let's say that the user enables the whole stack but doesn't use any config that depends on the guest components. In theory that could/should be non-fatal. My preference was actually to use rust features for all this because we can easily express dependency relationships, but we've moved away from build-time configuration.

Of course most of the time all the components will be there so idk how much this matters. In the other thread I suggested just logging the errors on startup and then failing later if someone tries to use a feature that is missing a dependency. I don't really have a preference, but either way it might be good to settle on some pattern that we can follow more generally.

Reply:
> For instance if you enable the spi-server-rest, you need the whole stack. Other features (will) only require the CDH/AA. I am not against adding a config to explicitly set which processes are running, but it makes the user have to do a little bit more work.

to simplify, the configuration here is incremental, so ASR implies CDH and AA, and CDH implies AA. in theory you could have ASR just to get evidence from AA without CDH, but I don't think we need to cover that. 

> but let's say that the user enables the whole stack but doesn't use any config that depends on the guest components. In theory that could/should be non-fatal

I think that's fair, but I wouldn't let processes fail to launch in cases of missing configuration but do the legwork and make sure that we're not launching them when we miss the prerequisites

Reply:
yeah, if we're introducing a config to explicitly specify which components are launched it probably is best to fail if we can't launch them. I guess the only drawback here is that the client may have to fiddle with this config in some cases.